### DISCUSSION TITLE
how to calculate unpolarized reflectance from PyElli Result

### DISCUSSION THREAD

**Discussion by paola-dececco** [PyEllips/pyElli#137] [Q&A] [answered]
Hello,

I wanted to use PyElli to analyze data from a normal incidence reflectometer.  None of the field returned in Result seem to contain unpolarized, normal reflectance (Result.R does not match the result from other solvers)

Help!

Paola

I tested it on a simple stack: Air - 20A,TaucLorentz oscillator sum - Si. below is the code

NSi=sidata['N']  # complex refractive index n+ik
Si_table = elli.Table(TLdata['Wvl'],NSi)
Layer = [elli.Layer(DM_SnO.get_mat(), 20)]
exp = elli.Experiment(elli.Structure(elli.AIR, Layer, Si_table.get_mat()),TLdata['Wvl'],0)
Result=exp.evaluate(solver=elli.Solver2x2) 
Rdata['PyElli Reflectance'] = Result.R
Rdata['PyElli WVL'] = exp.lbda

**Comment by domna** [comment]
Hey Paola,

at first glance everything looks fine to me, the `R` field is correct to get the reflectance of the material stack. We did, however, mainly benchmark against ellipsometry calculations, so it could also be that there is still a small error in the extraction of this value (Maybe @MarJMue has more experience with calculating reflectance and an example?). I will try to construct a small example later on to see whether we can reproduce literature results correctly.

If you are able and willing to share your model data (mainly the NSi an DM_Sn0 data) and if you have experimental data you want to compare to, then I'll have a look at this as well and see if I can spot any problems.

**Reply by domna** [reply]
I build a small example:
```
import elli
import numpy as np
import pandas as pd

lbda = np.linspace(300, 800, 500)
Si = elli.db.RII().get_mat("Si", "Aspnes")
SiO2 = elli.Cauchy(1.452,36.0).get_mat()
structure = elli.Structure(elli.AIR, [elli.Layer(SiO2, 400)], Si)
pd.DataFrame(
    {'refl': structure.evaluate(lbda, 0, solver=elli.Solver2x2).R}, 
    index=lbda
).plot(backend='plotly')
```

from which I get this result
<img width="668" alt="Bildschirmfoto 2023-04-20 um 07 33 53" src="https://user-images.githubusercontent.com/1014422/233267556-19c6b37b-2725-46f5-a6db-115025b93933.png">

which seems to roughly agree with this literature values
https://www.researchgate.net/figure/reflectivity-of-SiO2400nm-Si-structures-and-silica-spheres-deposited-on-SiO2400nm-Si_fig1_266403576

**Reply by paola-dececco** [reply]
Thank you for the fast response!

I found another legit difference between Py Elli and Wvase, but it still does not explain the difference in R. 
The difference is in the Tauc Lorentz parameterization choice (one Eg vs each oscillator its Eg)

I am using a off the shelf silicon 25C standard n,k in the 200-800nm region, 0.5nm spaced. I use the same dispersion on both program so you can pick any substrate as long as it is opaque.  I am putting on top a tauc lorentz with 3 oscillators params = ParamsHist()
params.add("SnO1_A", value=100, min=-100, max=100, vary=False)
params.add("SnO1_E", value=7, min=-40000, max=40000, vary=False)
params.add("SnO1_C", value=3, min=-40000, max=40000, vary=False)
params.add("SnO2_A", value=75, min=-100, max=100, vary=False)
params.add("SnO2_E", value=5, min=-40000, max=40000, vary=False)
params.add("SnO2_C", value=0.5, min=-40000, max=40000, vary=False)
params.add("SnO3_A", value=25, min=-100, max=100, vary=False)
params.add("SnO3_E", value=3, min=-40000, max=40000, vary=False)
params.add("SnO3_C", value=0.5, min=-40000, max=40000, vary=False)
params.add("Eg", value=0, min=0, max=40000, vary=False)
params.add("Einf", value=1.5, min=-40000, max=40000, vary=True)
#calculate the same TL in PyElli
DM_SnO=elli.TaucLorentz(Eg=params["Eg"])
DM_SnO=DM_SnO.add(A=params["SnO1_A"], E=params["SnO1_E"], C=params["SnO1_C"])
DM_SnO=DM_SnO.add(A=params["SnO2_A"], E=params["SnO2_E"], C=params["SnO2_C"])
DM_SnO=DM_SnO.add(A=params["SnO3_A"], E=params["SnO3_E"], C=params["SnO3_C"])

DM_SnO=elli.DispersionSum(DM_SnO, elli.EpsilonInf(eps=params['Einf']))

 PyElli uses a single Eg representation why my comparative program uses a parametrization with each Tauc-Lorentz having its own Eg. the weird part is that I thought as long as I set all Egs to be the same number [tested with Eg=2] e2 is going to be zero below that value, so they would be equivalent, and indeed n, k, e1,e2 between the programs agree to the 5th decimal digit. 

But R (unpolarized reflectance) is completely different (not just numerically) 

One more question : how do you handle the substrate? Do you stop at the reflection from the surface of the substrate when opaque? is there a switch between opaque and transparent that maybe I should set?

I am very thankful for your assistance. For fairness, although I am a scientist and strong believer in open source for code and for everything else, I do work for a private company. You can find my profile in Linkedin. Please let me know if this is not appropriate.

**Reply by domna** [reply]
Thank you for the model data. I don't have time right now to look into the details but will give it a try (hopefully) later today.

> Thank you for the fast response!
> 
😎
> PyElli uses a single Eg representation why my comparative program uses a parametrization with each Tauc-Lorentz having its own Eg. the weird part is that I thought as long as I set all Egs to be the same number [tested with Eg=2] e2 is going to be zero below that value, so they would be equivalent, and indeed n, k, e1,e2 between the programs agree to the 5th decimal digit.
> 
That's interesting. I observed the same strange behaviour with SpectraRay from Sentech. They also use a Tauc-Lorentz with an Eg associated with each oscillator. Back then I did some digging and found that in the literature there should only be one Eg for the material (which makes more sense to me in a physical sense as well). You can also refer to the original paper [here](https://pubs.aip.org/aip/apl/article/69/3/371/1023110/Parameterization-of-the-optical-functions-of), which we are using in pyElli. There is also a [whitepaper](https://pdf4pro.com/view/tauc-lorentz-dispersion-formula-horiba-4b026b.html) from Horiba which does it this way and with which I checked the pyElli model.

The thing is for SpectraRay it was not properly documented what they are doing internally so I did not implement their Tauc-Lorentz model, as in my view it is not correct (at least in the original sense of the paper). But since Woollam is doing the same there may be an alternative formulation. Do you have some documentation for the Tauc-Lorentz model you are using? If so I can add this to pyElli, too.

> But R (unpolarized reflectance) is completely different (not just numerically)

I see, can you share your wvase output so I can compare as soon as I test this model?

> One more question : how do you handle the substrate? Do you stop at the reflection from the surface of the substrate when opaque? is there a switch between opaque and transparent that maybe I should set?

The incident material (typically air) and substrate are dealt with as being infinite in size. It may be transparent in the calculation but we would not take any reflections coming back from a back surface into account. If you want to include something like this in your calculations you may use air as the substrate and add your substrate with a thickness to the layers.

> 
> I am very thankful for your assistance. For fairness, although I am a scientist and strong believer in open source for code and for everything else, I do work for a private company. You can find my profile in Linkedin. Please let me know if this is not appropriate.

No worries, this is totally fine. The software is open for researchers in private companies, too and you may also use it in your companies software according to our license (GPL-3.0).

**Reply by domna** [reply]
I now had time to have a look at this.
I reproduced your example and this is the curve I get when I extract `R` (with a Silicon substrate):

![Unknown](https://user-images.githubusercontent.com/1014422/233451024-e8b86ddd-bcd2-4ccf-b754-9d2279e87bc1.png)


And this is how the dielectric function of the SnO (generated with `DM_SnO.get_dielectric_df().plot()`) looks like:
![Unknown-1](https://user-images.githubusercontent.com/1014422/233451186-d9e0612b-daec-4c73-9c8a-337bbb3effe8.png)

Is this in accordance with your findings?

However, I think it would be helpful if you could share your reference data for this if this is possible (I don't have a wvase program available, so I cannot test it myself). Maybe there is some transformation of the data in wvase we do not do in pyElli, because as you explained the dielectric functions from wvase and from pyElli match as long as you set all Eg's in wvase to one fixed value, right? IMO the reflectance should also match in this case.

**Reply by paola-dececco** [reply]
I am putting together the data and we'll upload them soon. Sorry for the delay

**Reply by paola-dececco** [reply]
So as for the Eg [on for all oscillators or one per oscillators] I agree with you that one EG for all oscillators makes more sense. On the other end - Wvase offer both versions, and historically my company uses the one with many Eg, might be hard to convince people to move their baseline - even if it would make sense. I am attaching Wvase manual with details description page 8 - 23 talks about the 2 Tauc-Lorentz implementations 
[WVASE_Manual.pdf](https://github.com/PyEllips/pyElli/files/11291076/WVASE_Manual.pdf)
most likely I am making a mistake setting up PyElli - but if you think PyElli could be improved with access to WVase, I can generate any other test data you'd like me to and share it.

**Reply by domna** [reply]
Yes, I see. From how I understand it is that they are just adding different Tauc-Lorentz functions together with completely separated parameters. You can achieve this in pyElli by just adding completely separated Tauc-Lorentz dispersions with only one oscillator together, like so:

```
DM_SnO_multi = elli.TaucLorentz(Eg = params["SnO1_Eg"]).add(A=params["SnO1_A"], E=params["SnO1_E"], C=params["SnO1_C"])
DM_SnO_multi += elli.TaucLorentz(Eg = params["SnO2_Eg"]).add(A=params["SnO2_A"], E=params["SnO2_E"], C=params["SnO2_C"])
DM_SnO_multi += elli.TaucLorentz(Eg = params["SnO3_Eg"]).add(A=params["SnO3_A"], E=params["SnO3_E"], C=params["SnO3_C"])

DM_SnO_multi += params['Einf'].value
```
You may use this for testing now. If you find it useful we can add a `TaucLorentzMulti` dispersion or so.

I used this parameterset and get the exact same result from above:
```
params = ParamsHist()
params.add("SnO1_A", value=100, min=-100, max=100, vary=False)
params.add("SnO1_E", value=7, min=-40000, max=40000, vary=False)
params.add("SnO1_C", value=3, min=-40000, max=40000, vary=False)

params.add("SnO2_A", value=75, min=-100, max=100, vary=False)
params.add("SnO2_E", value=5, min=-40000, max=40000, vary=False)
params.add("SnO2_C", value=0.5, min=-40000, max=40000, vary=False)

params.add("SnO3_A", value=25, min=-100, max=100, vary=False)
params.add("SnO3_E", value=3, min=-40000, max=40000, vary=False)
params.add("SnO3_C", value=0.5, min=-40000, max=40000, vary=False)

params.add("Eg", value=0, min=0, max=40000, vary=False)
params.add("Einf", value=1.5, min=-40000, max=40000, vary=True)

params.add("SnO1_Eg", value=0, min=0, max=40000, vary=False)
params.add("SnO2_Eg", value=0, min=0, max=40000, vary=False)
params.add("SnO3_Eg", value=0, min=0, max=40000, vary=False)
```

**Reply by MarJMue** [reply]
Hi Paola,
as @domna mentioned we both have no access to woollam devices and it is always helpful if someone benchmarks our code agianst other software. So we are thankful for all requests like this. 
I will try to find out more, how WVASE calculates reflection.

**Comment by paola-dececco** [comment]
[ReflectanceComparison.zip](https://github.com/PyEllips/pyElli/files/11291032/ReflectanceComparison.zip)
Again sorry for the delay, I am not exactly the best python programmer - switch from MATlab just a month ago, and wanted to write a notebook to send you with data as well. In the zip folder you should find:
1. SiliconDM: standard Si dispersion table 200-800nm range
2. LorentzDM = Dispersion table generated by Wvase as 2 Lorentz oscillators with einf=1.5 in the 200-800nm range
3. Excel file with Reflectance calculated by Wvase for Thickness=20A (nominal), 2A, 68A, 134A and 200A for a film stack of Air/LorentzDM [Thickness]/SiliconDM

The notebook attempt to 
1. Compute in PyElli the same Lorentz dispersion - results are close but not exactly the same. Since e2 is the same, but e1 is slightly different I wonder if it is a Cauchy integral convergence issue
2. Compute in PyElli reflectance for all the thicknesses mentioned above and compare with WVase - here we are really far.

**Reply by domna** [reply]
Thank you, I'll try to reproduce everything when I have time and come back to you with my findings.

**Reply by domna** [reply]
Just for you to know. I'm working in this platform https://nomad-lab.eu/prod/v1/staging to work with the example (it is an open research data management platform - I am a developer at Humboldt University Berlin, where we develop this software). We have an online jupyterlab there with pyElli installed.

If you like you can create a free account and we can share an upload to work on your example together (which may be easier than exchanging code snippets).

**Reply by domna** [reply]
Hey @paola-dececco , I was able to execute your notebook and I see the difference you mentioned. I don't found an answer to the problem yet, but I also don't have too much time today to look into this. At first I thought the substrate may be different in pyElli but I compared it to one of our standard Si tables and they are pretty similar:
<img width="818" alt="Bildschirmfoto 2023-04-21 um 15 55 37" src="https://user-images.githubusercontent.com/1014422/233654426-b8ecc9c3-c0fa-49f4-a0ed-fe136672be13.png">

I currently don't know how this difference may be explained, but I will dig into it as soon as I have more time.

**Reply by paola-dececco** [reply]
Unfortunately my company 'security' does not allow me to access the NOMAD site. I will have to escalate a request for it - which will take few days...

**Reply by domna** [reply]
No worries, it's not a necessity. It just was convenient in another case we tried out some things. So feel free to join or not to join, whatever you prefer. It's also no problem for me to share code snippets with you.

**Reply by paola-dececco** [reply]
GOOD news!
I did find the discrepancy. your Lorentz energy assume einf=1. mine assume einf=0, so for eing=1.5 in PyElli I need to set einf=0.5, and it matches Wvase with Einf=1.5
So once again I can get the dispersion model to match exactly. 
Now the only difference is in the normal unpolarized reflectance... looking at the calcs Result.R=Rss/Rpp  I think what I need is sqrt(Rss *Rss_conj+Rpp*Rpp_conj) let me try that

**Reply by paola-dececco** [reply, marked as answer]
EVEN BETTER news!
I think all discrepancy is resolved. When I look into result.py to change the R definition - I realized it is the documentation which is confusing (Rss/Rpp) but the code is correct that is where it hit me... I work in A because the largest layer I see is <50A nut pyElli works... in nm! now everything matches perfectly. So I will have to be extra careful with the parameterization of dispersion model, or decide to add some to pyElli - let me know if you'd be interested in additional flavors of dispersion model if I go that route. Thank you Thank you!

**Reply by domna** [reply]
This is indeed great news 🎉

Yes, sorry for the wrong documentation. I spotted this as well while checking this but found the underlying code correct. Anyways, we will correct this (and thank you for spotting this!).
Indeed, we do use `einf=1` for the Lorentz functions. Maybe we should drop this to make the building of the models via summing cleaner and clearer.

Thank you for your feedback. It is very valuable for us to check this against other programs and to find common pitfalls while working with pyElli. It would be really nice and helpful if you let us now your experience with different models and what we could do to make it easier for people coming form other ellipsometry software. We definitely are interested in flavours of dispersion models.

**Comment by paola-dececco** [comment]
[Erratum_Jellison Modine Tauc Lorentz.pdf](https://github.com/PyEllips/pyElli/files/11291069/Erratum_Jellison.Modine.Tauc.Lorentz.pdf)
Maybe you are aware - in the original paper there was an error in the formula for the Tauc-Lorentz. probably not a problem in pYElli - sharing just in case

**Reply by domna** [reply]
Thank you. Yes we are aware of this Erratum and our model is build according to the corrected formula.
